### ISSUE TITLE
[1.1.4] asynch Scriptrunner initialization at startup needs some fine tuning

### ISSUE THREAD

**Issue by RaiMan** [RaiMan/SikuliX1#94]
reference: pull: https://github.com/RaiMan/SikuliX1/pull/90

**Comment by balmma** [comment]
Any objections against implementing lazy loading for script runners? This way we could get rid of the asynchronous init entirely. The required runner (and only the required one) gets just loaded before the first script run. This will also improve the startup time using the -r option.

Needs some minor regactoring in the ScriptingSupport class and the ScriptRunners.

If it is OK for you, I'll give it a try.

**Comment by RaiMan** [comment]
If you want, you can do that of course :-)

So I can concentrate on other things.

I would really love to see, that in the end, the stuff could be moved to the API.
Be aware that there are already redundancies.

**Comment by balmma** [comment]
Alright, lazy loading of runners in IDE is done.
Yes, moving all this stuff over to the API would be great.
Will concentrate on this during the next few days.

**Comment by balmma** [comment]
Was working the whole day on the refactoring of the runners and doing good progress. Since I mainly work on the API's Runner.java I would highly appreciate I you could resist doing too much heavy changes there. It just makes merging easier for me in the end :-)

**Comment by RaiMan** [comment]
Sorry and of course ok. Will not touch the Runner any more.

**Comment by balmma** [comment]
No worries, small changes as you've already done are really not a big issue :-)

**Comment by RaiMan** [comment]
I need a change to Runner:
``  public static File getScriptFile(File fScriptFolder)``
if it gets a file with the ending .py, it must return that immediately.

Background:
I am nearly through with supporting opening/editing/running plain .py files in the IDE. 
([the request on Launchpad](https://answers.launchpad.net/sikuli/+question/679034))
... only the -r option on commandline does not work.

**Comment by balmma** [comment]
Just go ahead. No problems.

**Comment by RaiMan** [comment]
ok. pushed.

**Comment by balmma** [comment]
I do great progress on this one.
If you want to have a preview you can have a look at https://github.com/SKOORAG/SikuliX1/tree/runners_to_api .
It already works running scripts from within the IDE. I'll probably need two more days to ensure everything else works as before.

**Comment by balmma** [comment]
And yes, it still needs a lot of cleanup :-)

**Comment by RaiMan** [comment]
Uuuups ;-) ... had a quick look online and now I KNOW (before I only could imagine ;-), why I always walked around this challenge.
I have downloaded and will look into it a bit deeper.
Tell me, if I can help something.

**Comment by balmma** [comment]
Refactoring is mostly done (your changes in the runner are already incorporated).

I've already tested running scripts from within the IDE, using the -r option and inside the scriptrun server.

The NetworkRunner seems to work as well:

`Runner.run("https://balmi.net/test/test.py")`
`Runner.run("https://balmi.net/test")`

I'm pretty sure that I broke some functionality and I would be very grateful if you could test the more specialized stuff before I rebase my branch and create the pull request.

**Comment by RaiMan** [comment]
Ok, I will try to do my very best ;-)
Work for tomorrow.

**Comment by RaiMan** [comment]
a first impression:
inserted remarks/headers start with ***... or +++...

```
+++++++++++++++++++++++ case startup
[3210 debug] Runner: initScriptingSupport: enter
[3554 debug] Runner: initScriptingSupport: added: jruby
[3555 debug] Runner: initScriptingSupport: added: Robot
[3556 debug] Runner: initScriptingSupport: added: Text
[3556 debug] Runner: initScriptingSupport: added: PackedSikulix
[3568 debug] Runner: initScriptingSupport: added: jython
[3568 debug] Runner: initScriptingSupport: added: NetworkRunner
[3568 debug] Runner: initScriptingSupport: added: PowerShell
[3569 debug] Runner: initScriptingSupport: added: AppleScript
[3569 debug] Runner: initScriptingSupport: added: git
Warning: Nashorn engine is planned to be removed from a future JDK release
[3927 debug] Runner: initScriptingSupport: added: JavaScript
[3928 debug] Runner: initScriptingSupport: added: Sikulix
[3928 debug] Runner: initScriptingSupport: warning: org.sikuli.script.runners.InvalidRunner.<init>()
[3928 debug] Runner: initScriptingSupport: exit with defaultrunner: jruby (rb)

**** default runner should be jython and same goes for type of an an Untitled tab (empty at startup)

++++++++++++++++++ case: image not found (FindFailed Exception)

********************************* output in terminal (option -c / -Dsikuli.console=false)

[debug] Region: handleFindFailed: img.png
[debug] Region: wait: img.png did not appear [3352 msec]
[error] script [ /Users/raimundhocke/Dropbox/SikuliX/Python/images.sikuli ] stopped with error in line 8
[error] FindFailed ( img.png: (110x39) in R[0,0 1280x800]@S(0) )
[debug] RunTimeIDE: ***** running cleanUp *****

********************************* IDE message area

[debug] Finder2: doFind: end 223 msec
[debug] Region: handleFindFailed: img.png
[debug] Region: wait: img.png did not appear [3138 msec]

[debug] Finder2: doFind: in original: %60,2995 (?90) 222 msec 
[debug] Finder2: doFind: end 223 msec
[debug] Region: handleFindFailed: img.png
[debug] Region: wait: img.png did not appear [3138 msec]
[error] script [ /Users/raimundhocke/Dropbox/SikuliX/Python/images.sikuli ] stopped with error in line 8
[error] FindFailed ( img.png: (110x39) in R[0,0 1280x800]@S(0) )
[debug] EditorConsolePane: redirection success for jython

[error] script [ /Users/raimundhocke/Dropbox/SikuliX/Python/images.sikuli ] stopped with error in line 8
[error] FindFailed ( img.png: (110x39) in R[0,0 1280x800]@S(0) )

[debug] EditorConsolePane: redirection success for jython
[debug] RunTimeIDE: ***** running cleanUp *****

[debug] RunTimeIDE: ***** running cleanUp *****

*** in IDE things seem to be doubled

*** in both cases the error line is not marked red in the IDE 

++++++++++++++++++++++ cases not yet tracked to the end

*** handling of an Untitled tab (not yet saved) does not look stable
    when script is saved and reloaded, the images are not found.

```

... good night ;-)

**Comment by balmma** [comment]
Allright, setting Jython as default is simple. What is the expected behaviour if Jython is not available?

**Comment by balmma** [comment]
I already have an idea about the double execution stuff and missing red lines. Will dig into this tomorrow.

**Comment by balmma** [comment]
Jython as Default is solves, double logs (in fact it was no double execution) as well. If Jython is not available, it falls back to JRuby.

Now I've to get rid of the global JythonHelper and JRubyHelper instances. They don't work very well if we instantiate multiple JythonRunners or JRubyRunners.

**Comment by balmma** [comment]
Additionally I had to synchronize access to JythonHelper and JrubyHelper quite a bit. Those two are most probably my next refactoring targets :-)

**Comment by balmma** [comment]
Stuff from upstream/master is also merged in.

**Comment by RaiMan** [comment]
**default scripting context**
 - Jython if on classpath
 - if not, then JRuby, if on classpath
 - if not, then JavaScript

**Comment by balmma** [comment]
> **default scripting context**
> 
> * Jython if on classpath
> * if not, then JRuby, if on classpath
> * if not, then JavaScript

Done

**Comment by RaiMan** [comment]
Ok, thanks. I will play around with it now.

**Comment by RaiMan** [comment]
JythonRunner::doInit is run with every scriptrun - should only be run the first time

**Comment by balmma** [comment]
It is already implemented this way in the AbstractScriptRunner. The problem here was that SikulixIDE was closing the current runner after every run. This was already the case before the refactoring. Close call is removed now.

**Comment by RaiMan** [comment]
I have now cloned your repo branch. 

In parallel to your work I added some features to the IDE (mainly opening/running plain Python files and open/edit/save text files).

I already found some quirks related to these new features, that I would like to eval and fix in your branch. 
Can you add me as a member, so I can push my commits?

**Comment by balmma** [comment]
You should have push permissions now.

**Comment by RaiMan** [comment]
yes, thanks - works.

**Comment by RaiMan** [comment]
Had to make a fix to the master.
Not sure, how to get this into the branch runners_to_api now :-(

**Comment by balmma** [comment]
ok, I'll have it rebased in a few minutes.

**Comment by RaiMan** [comment]
ok, worked.

**Comment by balmma** [comment]
I moved all the default runner stuff from the API to the IDE. It just makes more sense this way. Why should the API care about IDE default runners?

**Comment by RaiMan** [comment]
Makes sense. Will have a look later today.

**Comment by RaiMan** [comment]
LOL, you silently changed my "implementation" to select runners representing currently supported IDE-tab content. I already thought, Git is not liking me, because my isIdeContent() vanished ;-)

Ok, agreed. The runner reference is its class and additional attributes should be implemented in ScriptingSupport, as long as no additional Runner features are needed.

**Comment by balmma** [comment]
Agree, should have mentioned that :-)

**Comment by RaiMan** [comment]
No problem at all. With a cup of coffee ...

**Comment by balmma** [comment]
> The runner reference is its class and additional attributes should be implemented in ScriptingSupport, as long as no additional Runner features are needed.

Everything that is IDE specific and the API shouldn't care about should be implemented in ScriptingSupport. In this case the API simply shouldn't know which runners are supported by the IDE.

**Comment by RaiMan** [comment]
ok, thanks. Understood and agreed.

**Comment by balmma** [comment]
What is the current testing status BTW?

**Comment by RaiMan** [comment]
I am on the way, more time for that later today.

Line Error mark is not yet back. Just say yes and I will try to fix it later.

**Comment by RaiMan** [comment]
... but I think, latest this evening you can create a pull request.

**Comment by balmma** [comment]
I've already found the reason for this :-) Now I now what the following line in the Jython runner is for:

>  if (isFromIDE) {
          exitCode *= -1;
        } else {
          exitCode = 1;
        }

**Comment by RaiMan** [comment]
yes, also arrived there and found, that beginning from SikuliIDE::ButtonRun the info, that the script run was called from inside the IDE has been stripped.
I remember, that some years ago I inserted some tricky things, that are not obvious - sorry ;-)

TBD: the runners should know somehow, that they are called from the IDE (parameter options()?), so they return appropriate.

But as said: I can take over to fix it.

**Comment by balmma** [comment]
Yes, the options might be a good place. But I would rather call the option like "returnErrorLine" instead of  "runForIDE".

**Comment by balmma** [comment]
Your can decide if you want to fix it yourself or if should do it :-)

**Comment by RaiMan** [comment]
Ok, will do :-)

**Comment by RaiMan** [comment]
... more complicated than expected :-(

 - Jython seems to work
 - JRuby has to be fixed (error handling)
 - JavaScript not yet touched (does not work anyways ;-)

irritating (thought that the runner init was deferred in IDE too):
added some Debug.info() and get this at IDE startup with no preloaded scripts: 
```
Warning: Nashorn engine is planned to be removed from a future JDK release
[info] Jython init
[info] Jython redirect
[info] JRuby init
[info] JRuby redirect
[info] JavaScript init
Warning: Nashorn engine is planned to be removed from a future JDK release
[info] Jython init
```
Would have expected, that this comes only at first script run and only for one runner.

**Comment by balmma** [comment]
No, we need to init the runner before we can do the IO redirection. And the double Jython init comes due to the fact that RobotRunner extends the JythonRunner. You should use NAME or getName in your debug logs :-)

**Comment by balmma** [comment]
Probably it would anyway be better to just remember the pipes in  AbstractScriptRunner.redirect and apply it upon init using doRedirect() . If the runner has already been initalized, doRedirect() can be called directly. This way we could also get rid of the async execution in ConsolePane.

**Comment by balmma** [comment]
Are you done with your work for today and have everything pushed? Then I'll optimize this redirection stuff this evening.

**Comment by balmma** [comment]
Wow, this whole thing is going to be better and better :-)

**Comment by RaiMan** [comment]
ok, understood and agreed. Please just remove the stupid Debug.info() ;-)

I am finished for today and all is pushed.

tomorrow: JRuby, JavaScript, Robot and a look over all the other run usages

good night

**Comment by balmma** [comment]
If I were you I wouldn't invest too much time on the JavaScript ttuff. I'm pretty shure that  we have to use another engine there anyway. E.g. node.js through some JNI bridge like Node4J or similar.

**Comment by RaiMan** [comment]
JavaScript: I think I should just switch it off for now, since it is not really useable anyways.

When I am through with the support for using SikuliX from real Python, I might come back to that.

**Comment by balmma** [comment]
Yes, then instead of investing time on it just  turn it off (using it's new isSupported method of course :-) )

**Comment by balmma** [comment]
Fixed things:
- Do not load runner anymore on redirect but memorize pipes instead. Apply them as soon as runner gets initialized
- Pre initalize current tab runner on IDE start to speedup first script run
- Nicer options for runner (IScriptRunner.Options). Overall less shaky :-)
- Make JRuby error line handling working
- Make JavaScript working and add very very basic error line handling

**Comment by balmma** [comment]
BTW: the title of this issue is a massive understatement :-)

**Comment by RaiMan** [comment]
very good job :-) thanks.

The IScriptRunner.Options is a good, flexible solution. I will move the execBefore and exceAfter logic there next week.

JavaScript: I will set it unsupported and revise it a little bit the next weeks.

If you are done for now, then I think you should create a pull request, to make it available to the world.

**Comment by RaiMan** [comment]
> BTW: the title of this issue is a massive understatement :-)

feel free to change it to what ever you think is appropriate or make a suggestion ;-)

**Comment by RaiMan** [comment]
I am done for today ;-)
Wish you a nice weekend.
Great thanks again for your great contributions. I learned a lot :-)

**Comment by RaiMan** [comment]
The enhancements seem to be stable now.

**Comment by balmma** [comment]
Glad to hear that :-)
